northd/multicast.dl

```diff
@@ -0,0 +1,161 @@
+                              mCAST_DEFAULT_MIN_QUERY_INTERVAL_S(),
+                              mCAST_DEFAULT_MAX_QUERY_INTERVAL_S()),
+    var query_max_resp =
+        map_get_int_def(other_config, "mcast_query_max_response",
+                        mCAST_DEFAULT_QUERY_MAX_RESPONSE_S()).
+
+/* IP Multicast per router configuration. */
+relation &McastRouterCfg(
+    datapath: uuid,
+    relay   : bool
+)
+
+&McastRouterCfg(lr_uuid, mcast_relay) :-
+    nb.Logical_Router(._uuid = lr_uuid, .options = options),
+    var mcast_relay = map_get_bool_def(options, "mcast_relay", false).
+
+/* Mapping between Switch and the set of router port uuids on which to flood
+ * IP multicast for relay.
+ */
+relation SwitchMcastFloodRelayPorts(sw: Ref<Switch>, ports: Set<uuid>)
+
+SwitchMcastFloodRelayPorts(switch, relay_ports) :-
+    &SwitchPort(
+        .lsp  = lsp,
+        .sw = switch,
+        .peer = Some{&RouterPort{.router = &Router{.mcast_cfg = &mcast_cfg}}}
+    ), mcast_cfg.relay,
+    var lsp_uuid = lsp._uuid,
+    var relay_ports = Aggregate((switch), group2set(lsp_uuid)).
+
+SwitchMcastFloodRelayPorts(switch, set_empty()) :-
+    Switch[switch].
+
+/* Relation used to determine if a Switch has at least one router port
+ * connected to a Router with multicast relay enabled.
+ */
+relation SwitchHasMcastFloodRelay(sw: Ref<Switch>, has_flood_relay: bool)
+
+SwitchHasMcastFloodRelay(switch, not set_is_empty(ports)) :-
+    SwitchMcastFloodRelayPorts(switch, ports).
+
+/* Flattened IGMP group. One record per address-port tuple. */
+relation IgmpSwitchGroupPort(
+    address: string,
+    switch : Ref<Switch>,
+    port   : uuid
+)
+
+IgmpSwitchGroupPort(address, switch, lsp._uuid) :-
+    sb.IGMP_Group(.address = address, .datapath = igmp_dp_set,
+                  .ports = pb_ports),
+    var pb_port_uuid = FlatMap(pb_ports),
+    port in &SwitchPort(.lsp = lsp, .sw = switch),
+    sb.Port_Binding(._uuid = pb_port_uuid, .logical_port = lsp.name).
+
+/* Aggregated IGMP group: merges all IgmpSwitchGroupPort for a given
+ * address-switch tuple from all chassis.
+ */
+relation IgmpSwitchMulticastGroup(
+    address: string,
+    switch : Ref<Switch>,
+    ports  : Set<uuid>
+)
+
+IgmpSwitchMulticastGroup(address, switch, ports) :-
+    IgmpSwitchGroupPort(address, switch, port),
+    var ports = Aggregate((address, switch), group2set(port)).
+
+/* Flattened IGMP group representation for routers with relay enabled. One
+ * record per address-port tuple for all IGMP groups learned by switches
+ * connected to the router.
+ */
+relation IgmpRouterGroupPort(
+    address: string,
+    router : Ref<Router>,
+    port   : uuid
+)
+
+IgmpRouterGroupPort(address, rtr_port.router, rtr_port.lrp._uuid) :-
+    SwitchMcastFloodRelayPorts(switch, sw_flood_ports),
+    IgmpSwitchMulticastGroup(address, switch, _),
+    var flood_port = FlatMap(sw_flood_ports),
+    &SwitchPort(.lsp = lsp, .peer = Some{&rtr_port}),
```

Comment:
> You can simplify this slightly:
> 
> ```
> &SwitchPort(.lsp = nb.Logical_SwitchPort{._uuid == flood_port}, .peer = Some{&rtr_port}).
> ```
> 
> More importantly, this will perform better, as it will use a version of `SwitchPort` indexed by `lsp._uuid`. (The compiler is not yet smart enough to figure this out from the original rule).

Interesting! The only reason I kept it as a different clause is that the line was getting too long and I'm never sure how to split the arguments of the relations in a uniform way. Fixed.